solr/core/src/java/org/apache/solr/search/SolrIndexSearcher.java

```diff
@@ -1712,33 +1806,47 @@ public ScoreMode scoreMode() {
       // no docs on this page, so cursor doesn't change
       qr.setNextCursorMark(cmd.getCursorMark());
     } else {
-
-      final TopDocsCollector topCollector = buildTopDocsCollector(len, cmd);
-      DocSetCollector setCollector = new DocSetCollector(maxDoc);
-      MaxScoreCollector maxScoreCollector = null;
-      List<Collector> collectors = new ArrayList<>(Arrays.asList(topCollector, setCollector));
-
-      if ((cmd.getFlags() & GET_SCORES) != 0) {
-        maxScoreCollector = new MaxScoreCollector();
-        collectors.add(maxScoreCollector);
+      TopDocs topDocs;
+      if (pf.postFilter != null || cmd.getSegmentTerminateEarly() || cmd.getTimeAllowed() > 0
+          || query instanceof RankQuery || query instanceof GraphQuery) {
```

Comment:
What makes GraphQuery so special?  Maybe we should define a marker interface here?  What if the Query is a BooleanQuery that somewhere inside _contains_ a GraphQuery?

Reply:
> What if the Query is a BooleanQuery that somewhere inside _contains_ a GraphQuery?

`QueryVisitor` to rescue.

Reply:
I should have clarified that I knew the answer is QueryVisitor :-). I wanted those working on the PR to remember that in general it's questionable to aQuery instanceof XYZ.